proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+  --      Probable fix: use a type annotation to specify what ‘a0’ should be.
+  --      These potential instance exist:
+  --        instance [safe] Sized Bool -- Defined at <interactive>:15:10
+  --    • In the first argument of ‘print_int’, namely ‘sizeOf’
+  --      In the expression: print_int sizeOf
+  --      In an equation for ‘main’: main = print_int sizeOf
+  --
+  main = print_int sizeOf
+
+It also means that eta-reduction is not possible::
+
+  -- Valid code:
+  mySizeOf :: forall a. Sized a => Int
+  mySizeOf @a = sizeOf @a
+
+  -- Eta-reduction attempt fails:
+  --
+  --  • Could not deduce (Sized a0) arising from a use of ‘sizeOf’
+  --    from the context: Sized a
+  --      bound by the type signature for:
+  --                 mySizeOf :: forall a. Sized a => Int
+  --    The type variable ‘a0’ is ambiguous
+  --
+  mySizeOf :: forall a. Sized a => Int
+  mySizeOf = sizeOf
+
+
+If we had visible ``forall``, for which there is already precedent at the
+type-level, we could design an API for ``sizeOf`` that has none of the issues
+listed above::
+
+  sizeOf :: forall a -> Sized a => Int
+
+This type captures the intent behind this function, and, if we allow it, its
+use would have the least noise and good error messages::
+
+  print_int :: Int -> IO ()
+
+  -- Valid code:
+  main = print_int (sizeOf Bool)   -- NB: no visibility override '@'
+
+
+  -- The parameter is not specified, good error message:
+  --
+  --    • Couldn't match expected type ‘Int’
+  --                with actual type ‘forall a -> Sized a => Int’
+  --    • Probable cause: ‘sizeOf’ is applied to too few arguments
+  --      In the first argument of ‘print_int’, namely ‘sizeOf’
+  --      In the expression: print_int sizeOf
+  --      In an equation for ‘main’: main = print_int sizeOf
+  --
+  main = print_int sizeOf
+
+Eta-reduction is now possible::
+
+  -- Valid code:
+  mySizeOf :: forall a -> Sized a => Int
+  mySizeOf a = sizeOf a
+
+  -- Eta-reduction attempt succeeds:
+  mySizeOf :: forall a -> Sized a => Int
+  mySizeOf = sizeOf
+
+The proposed visible ``forall`` would be an irrelevant quantifier. However, if
+we were to make it relevant, we would get full-blown dependent functions
+(pi-types). Therefore, implementing this feature would pave the road for future
+work on Dependent Haskell.
+
+To summarize, there are three reasons to make this change:
+
+* Language consistency (symmetry between terms and types)
+* Ability to design better APIs (good error messages, no proxy types, no ambiguous types)
+* Prepare the compiler internals for further work on dependent types
+
+Proposed Change Specification
+-----------------------------
+
+1. Add a new language extension, ``RequiredTypeArguments``.
+
+2. When ``RequiredTypeArguments`` is in effect, lift the restriction that the
+   ``forall a ->`` quantifier cannot be used in types of terms.
+
+3. In types of terms, ``forall a ->`` is an irrelevant quantifier.
```

Comment:
Yes. `forall a. t` is also relevant in kinds, irrelevant in types. So, although the difference in relevance is unfortunate, it has a precedent, which we adhere to. (Eventually we will want to sort all of this out, of course, but not now)

Reply:
Ah, that's quite fair then. Please mention this fact in the proposal itself. It's very helpful in judging the merit of the specification (the best place is probably right here, to be honest, but it could otherwise be in the effects and interaction section).

Reply:
The sentence is now:

> In types of terms, `forall a ->` is an irrelevant quantifier (as is `forall a.`).